src/ripple/shamap/FullBelowCache.h

```diff
@@ -114,38 +457,74 @@ class BasicFullBelowCache
     void
     insert(key_type const& key)
     {
-        m_cache.insert(key);
+        if (auto c = cache_.load())
+            c->insert(key);
     }
 
-    /** generation determines whether cached entry is valid */
+    /** Returns the generation that can determine if a cached entry is valid. */
     std::uint32_t
-    getGeneration(void) const
+    getGeneration() const
     {
-        return m_gen;
+        return generation_;
     }
 
+    /** Clears the cache.
+
+        This effectively replaces the cache with an entirely new instance, and
+        destroys the old instance.
+
+        It is safe to call this method from any thread.
+     */
     void
     clear()
     {
-        m_cache.clear();
-        ++m_gen;
+        auto c = cache_.load();
+
+        ++generation_;
+
+        cache_.store(std::make_shared<KeyCache>(
+            "FullBelow: " + name_,
+            targetSize_,
+            expiration_,
+            clock_,
+            journal_,
+            collector_));
+
+        if (c)
+        {
+            std::thread t(
+                [](std::shared_ptr<KeyCache> cache) {
+                    // just invoke the destructor of cache
+                },
+                std::move(c));
+            t.detach();
+        }
     }
 
-    void
-    reset()
+    Json::Value
+    info() const
     {
-        m_cache.clear();
-        m_gen = 1;
+        auto c = cache_.load();
+
+        while (!c)
+            c = cache_.load();
+
+        auto ret = c->info();
+        ret["generation"] = std::to_string(generation_);
+        return ret;
     }
 
 private:
-    CacheType m_cache;
-    std::atomic<std::uint32_t> m_gen;
-};
+    std::string const name_;
+    clock_type& clock_;
+    beast::Journal journal_;
+    beast::insight::Collector::ptr collector_;
+    std::size_t const targetSize_;
+    std::chrono::seconds const expiration_;
 
-}  // namespace detail
-
-using FullBelowCache = detail::BasicFullBelowCache;
+    std::atomic<std::uint32_t> generation_ = 0;
+    std::atomic<std::shared_ptr<KeyCache>> cache_;
```

Comment:
This: https://github.com/HowardHinnant/rippled/commit/ef666ae4951ba30ca521109defc5b5a59ef172ff must be merged in to get things to compile on the latest macOS tools.  Otherwise:
```
error: _Atomic cannot be applied to type 'std::shared_ptr<ripple::detail::FullBelowCacheImpl<ripple::base_uint<256>>>' which is not trivially copyable
  _LIBCPP_DISABLE_EXTENSION_WARNING _Atomic(_Tp) __a_value;
                                    ^
```